### DISCUSSION TITLE
new Audio() constructor

### DISCUSSION THREAD

**Discussion by faveoled** [scala-js/scala-js-dom#758] [Q&A]
https://developer.mozilla.org/en-US/docs/Web/API/HTMLAudioElement/Audio
```
new Audio()
new Audio(url)
```
How do I translate this to Scala.js DOM?

**Comment by armanbilge** [comment]
Hmm, I'm confused how this works since the name used by the constructor is different from the class. Good question! @sjrd can you enlighten us please 😅 🙏

**Reply by sjrd** [reply]
`(new Audio()) instanceof HTMLAudioElement === true`, so `Audio` seems to be a subclass of `HTMLAudioElement`.

**Reply by armanbilge** [reply]
Thanks! @faveoled would you like to open a PR to add a new `class Audio extends HTMLAudioElement` with the appropriate constructors?

**Reply by faveoled** [reply]
I'm not sure. Is there a similar class? What kind of package should it go to?

**Reply by armanbilge** [reply]
It can go in this package, check any of the others for examples. Feel free to open a draft PR and we can help :)

https://github.com/scala-js/scala-js-dom/tree/main/dom/src/main/scala/org/scalajs/dom

**Reply by faveoled** [reply]
ok, maybe I'll try tomorrow

**Reply by sjrd** [reply]
It reminds me of `Image`, which creates an `HTMLImageElement`.

**Reply by armanbilge** [reply]
Ah so it does! We are missing that one too 😅 
- https://github.com/scala-js/scala-js-dom/issues/621

**Reply by zetashift** [reply]
The PR in that one is in the bikeshedding stage :P

**Reply by armanbilge** [reply]
Oh I forgot 😂 lets do something about that.

**Reply by faveoled** [reply]
```scala
import scala.scalajs.js
import org.scalajs.dom.HTMLAudioElement
import scala.scalajs.js.annotation.JSGlobal

@JSGlobal("Audio")
@js.native
class Audio () extends HTMLAudioElement {
  def this(url: String) = this()
}
```
Tried this but it wasn't compiling so I gave up

**Reply by faveoled** [reply]
Found a possible implementation here
https://github.com/tmtsoftware/tmt-typed/blob/fdf18ba48c89466c7197e40f6d5c9b49b30e505f/modules/std/src/main/scala/typings/std/Audio.scala

**Reply by armanbilge** [reply]
> Tried this but it wasn't compiling so I gave up

What was the compile error?

**Reply by armanbilge** [reply]
Actually feel free to open a PR with that, so we can help more easily :)

**Reply by faveoled** [reply]
> > Tried this but it wasn't compiling so I gave up
> 
> What was the compile error?

```
class Audio needs to be abstract, since var srcObject: 
  scala.scalajs.js.UndefOr[org.scalajs.dom.MediaStream | 
    org.scalajs.dom.MediaSource
   | org.scalajs.dom.Blob] in class HTMLMediaElement in package org.scalajs.dom is not defined 
(Note that variables need to be initialized to be defined)
```

**Reply by armanbilge** [reply]
Woops, it looks like we forgot to write `= js.native` here:
https://github.com/scala-js/scala-js-dom/blob/260fdf00f106ef913871ba48d43196d4b69e8824/dom/src/main/scala/org/scalajs/dom/HTMLMediaElement.scala#L129-L130

**Comment by faveoled** [comment]
My solution so far is this:
```scala
import org.scalajs.dom.HTMLAudioElement
import scala.scalajs.js

def audio(url: String): HTMLAudioElement = {
  js.Dynamic
    .newInstance(js.Dynamic.global.Audio)(url)
    .asInstanceOf[HTMLAudioElement]
}

def audio(): HTMLAudioElement = {
  js.Dynamic
    .newInstance(js.Dynamic.global.Audio)()
    .asInstanceOf[HTMLAudioElement]
}
```

**Comment by zetashift** [comment]
I think you'll want something similiar to the `Image` PR, and feel free to open up a Pull Request! The steps to open one up are outlined here: https://github.com/scala-js/scala-js-dom/blob/main/CONTRIBUTING.md

```scala

class Image(width: Int, height: Int) extends HTMLImageElement {
  def this(width: Int) = this(width, 0)
  def this() = this(0, 0)
}
```
So maybe something like this( I am unsure how scala-js-dom deals with optional constructor values)

```scala
class Audio(url: js.UndefOr[String] = js.native) extends HTMLAudioElement {}
```

@armanbilge am I close with this?

**Reply by armanbilge** [reply]
That seems fine, you can also do two constructors instead of an `UndefOr` and default parameter, that tends to work better for type inference.

**Reply by zetashift** [reply]
```scala
class Audio(url: String) extends HTMLAudioElement {
  def this(url: String) = js.native
  def this() = js.native
}
```

I was unsure what the body of the constructors should be.

**Reply by armanbilge** [reply]
Maybe like this. Implementations don't matter (they are discarded), since its just a facade.

```scala
class Audio extends HTMLAudioElement {
  def this(url: String) = this()
}
```

**Reply by zetashift** [reply]
Right I forgot that!

so @faveoled if you have the time you could do the following

- Create a new `Audio.scala` in `./dom/src/main/scala/org/scalajs/dom/`
- The code for your Audio constructur will look like this:
```
class Audio extends HTMLAudioElement {
  def this(url: String) = this()
}
```
- run `sbt prePR` (this will take a while)
- commit the changes(the ones in `api-reports` folders as well)
 
If you need any help let me know!
